### DISCUSSION TITLE
Sheet: Issues with editing a newly added row.

### DISCUSSION THREAD

**Discussion by TheMightyE** [primefaces/community#1949] [PrimeFaces]
With the PrimeFaces Extensions sheet component, I have an option in the context menu to add a row above. When I select that option, I also add a row in that index in the list in the backing bean and the mock "database". The issue is when I edit this newly added row. In the `onCellChange` function in the `LibraryBean` the `sheet`'s rowMap still only has the three rows from when the sheet was first rendered, so the function ends up editing the incorrect row as it doesn't even know about the new row. I am trying to avoid updating the sheet after adding the new row for better performance, but is that truly my only option?

Here's the Quarkus project. Run it with `.\mvnw.cmd compile quarkus:dev`. If you use VS Code, you can debug the project with the `Debug Quarkus application` configuration.
[quarkus-sheet-add-row-issue.zip](https://github.com/primefaces/community/files/15342145/quarkus-sheet-add-row-issue.zip)

PrimeFaces Extensions v13.0.8

**Comment by melloware** [comment]
```
[INFO] No sources to compile
Error occurred during initialization of VM
Could not find agent library instrument on the library path, with error: Can't find dependent libraries
Module java.instrument may be missing from runtime image.
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  8.874 s
[INFO] Finished at: 2024-05-17T17:31:31-04:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal io.quarkus.platform:quarkus-maven-plugin:3.8.4:dev (default-cli) on project code-with-quarkus: Failed to run: Dev mode process did not complete successfully -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
```

**Reply by TheMightyE** [reply]
Hmm okay I'll try to figure this out. Are you just running `.\mvnw.cmd compile quarkus:dev` from the command line?

**Reply by melloware** [reply]
yep

**Reply by TheMightyE** [reply]
Not familiar with this issue. Is your JAVA_HOME env variable set? Mine is set to java 17. I also have path_to_java_17/bin as the first in my system PATH variable.

**Reply by TheMightyE** [reply]
I converted the quarkus project to the PrimeFaces Test jetty project. You can clone it [here](https://github.com/TheMightyE/primefaces-test-sheet-context-menu-bug.git) and run it with `.\mvnw.cmd clean jetty:run`.
Note: I was unable to import `test.js` so I embedded it in the xhtml.

I was unable to recreate the error you are getting with my quarkus project sadly.

**Reply by melloware** [reply]
here is my `mvn --version`

```
Apache Maven 3.9.5 (57804ffe001d7215b5e7bcb531cf83df38f93546)
Maven home: C:\tools\apache-maven-3.9.5
Java version: 17.0.6, vendor: Eclipse Adoptium, runtime: C:\tools\jdk-17.0.6+10
Default locale: en_US, platform encoding: Cp1252
OS name: "windows 11", version: "10.0", arch: "amd64", family: "windows"
```

**Reply by TheMightyE** [reply]
Okay my output of `.\mvnw.cmd -version` is below, though I wouldn't think this should be the issue.

```
Maven home: C:\...\.m2\wrapper\dists\apache-maven-3.9.6-bin\3311e1d4\apache-maven-3.9.6
Java version: 17.0.7, vendor: Amazon.com Inc., runtime: C:\...\jdk17.0.7_7
Default locale: en_US, platform encoding: Cp1252
OS name: "windows 11", version: "10.0", arch: "amd64", family: "windows"
```
Can you try to use the PrimeFaces Test project instead. My previous link takes you to master branch but this one should take you to the other correct branch: https://github.com/TheMightyE/primefaces-test-sheet-context-menu-bug/tree/sheet-edit-new-row-issue

**Reply by melloware** [reply]
I am not sure but i think this is your bug: `this.id = UUID.randomUUID().toString();`

Your `rowKey="#{row.id}"` means that is how the sheet is identifiying rows etc.

the reason I think that is in the network traffic F12 I can see your ID's changing when I do "Fetch Latest".

![image](https://github.com/primefaces/community/assets/4399574/1ab9b1d3-c096-436d-a2c7-98435afe743e)

CAVEAT: I did not deep dive on this just felt like this was "a smell".

**Reply by TheMightyE** [reply]
I see what you mean. However, it is okay if the ids are changing when I fetch the latest. It's important for them to not change between adding a new row and then editing it without fetching latest or reloading the page. Which they do not so that's fine.

**Comment by TheMightyE** [comment]
I spent more time on this, but didn't get far. For some reason, the SheetEvent in `onCellChange(final SheetEvent event)` doesn't have the updated data. It's always missing the newly inserted row. The handsontable does have the new row. I verified with `PF("sheetWidget").ht.getRowData()` in the browser console. Any ideas?

**Reply by melloware** [reply]
you might have to dig into how `onCellChange` is fired in the PFE code.
